### PR TITLE
Check Chainstart Before Processing Messages In Pubsub

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature Addition

**What does this PR do? Why is it needed?**

 - [x] Prevents gossip messages from being relayed before chainstart.
 - [x] Add unit test for gossip.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/sync/service.go [hunk 1]
```diff
@@ -277,6 +277,11 @@ func (s *Service) registerHandlers() {
 	}
 }
 
+// marks the chain as having started.
+func (s *Service) markForChainStart() {
```
Comment (terencechain) [inline beacon-chain/sync/service.go:281]:
```suggestion
func (s *Service) setChainStart() {
```

Sounds better?

File: beacon-chain/sync/service.go [hunk 2]
```diff
@@ -277,6 +277,11 @@ func (s *Service) registerHandlers() {
...
+	s.chainStarted = true
```
Comment (terencechain) [inline beacon-chain/sync/service.go:282]: Do we want to add a lock ?

Comment (nisdas) [reply beacon-chain/sync/service.go:282]: This method can only be called once and that too from a single routine. I don't think a lock is necessary here

**Comment by prestonvanloon** [comment]
@nisdas there might be some overlap here with #7496. Let me know if you think so and how to proceed with these two PRs.

**Comment by nisdas** [comment]
This PR mostly does a defensive check during genesis to make sure that we dont receive any messages before chainstart. It's different from #7496, which does a specific subscription check for each topic. @prestonvanloon

**Review by rauljordan — APPROVED** [review]

File: beacon-chain/sync/subscriber.go [hunk 3]
```diff
@@ -157,13 +157,18 @@ func (s *Service) subscribeWithBase(base proto.Message, topic string, validator
...
 	return topic, func(ctx context.Context, pid peer.ID, msg *pubsub.Message) (res pubsub.ValidationResult) {
 		defer messagehandler.HandlePanic(ctx, msg)
 		res = pubsub.ValidationIgnore // Default: ignore any message that panics.
 		ctx, cancel := context.WithTimeout(ctx, pubsubMessageTimeout)
 		defer cancel()
 		messageReceivedCounter.WithLabelValues(topic).Inc()
+		// Reject any messages received before chainstart.
+		if !s.chainStarted {
+			messageFailedValidationCounter.WithLabelValues(topic).Inc()
+			return pubsub.ValidationReject
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber.go:170]: I think this is not a good choice. ValidationReject affects peer scoring, we should use ValidationIgnore.

**Pull request merged**